crypto/src/key/hdkd/derivation_path.rs

```diff
@@ -26,6 +26,14 @@ const SEPARATOR: &str = "/";
 #[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
 pub struct DerivationPath(Vec<ChildNumber>);
 
+impl DerivationPath {
+    pub fn extend(&self, num: ChildNumber) -> Self {
+        let mut new_path = self.0.clone();
+        new_path.push(num);
+        new_path.into()
+    }
```

Comment:
I would rather call this join than extend. Both Rust and Python call it join. C++ calls it append.

Also one nice thing boost does is using the operator `/` for these kinds of extensions... it's kind of weird, but shows nicely in code. So if `a`, `b` and `c` are paths, we write in code:

```
let d = a / b / c;
```

In fact, the C++ standard adopted it since C++17: https://en.cppreference.com/w/cpp/filesystem/path/append

I'm not suggesting we go that far, but it's an idea worth thinking about if these joins become annoying or something. If I see it, I won't fight it.

Reply:
Join and append differ in that the RHS can have multiple components whereas the method presented here only adds one component at the end. In that respect, it's more similar to `Vec::push`. If it was mutating the argument in place, it would make sense to just call it `push`.

Since it is internally and conceptually a `Vec`, it makes sense to me for the API to mimic `Vec`. Maybe it is sensible to change this method to use `&mut self` and call it `push`. If you look how it's implemented, there is hidden clone followed by a push anyway. For completenes can also add `FromIterator` to add multiple child numbers in one go and `join` to concatenate two paths together. I think this would make the API largely unsurprising.

Reply:
Structurally, it's a path. But it's not implemented like a path in the API for some reason. If I understand correctly, you're not happy calling this function a join because the API we created doesn't act like a path. Well, I'd call that a problem in the API, not in the idea, since this structure is basically a path in every way. I hope I'm not missing something in this story.

And just to not give the vibe that I'm insistent, I wanna emphasize that I don't care too much about this detail since this is some esoteric thing that'll barely be used in the code. But it'd be nice to have something consistent coming out of it.

Reply:
I have changed the API by adding a push & pop operation.